tor/helpers/threaded_worker.py

```diff
@@ -0,0 +1,114 @@
+        """
+        return (
+            '0.1.0.ToR.Client.Thread.{}.ID.{} (contact u/itsthejoker)'.format(
+                random.randrange(0, 30),
+                ''.join(
+                    [random.choice(string.ascii_lowercase) for _ in range(6)]
+                ),
+            )
+        )
+
+    def parse_json_posts(posts: Dict) -> List:
+        trimmed_links = list()
+        number_of_posts = 10
+        posts_raw = posts['data']['children']
+        posts_raw = posts_raw[:number_of_posts]  # cut the list from 25 to 10
+        for item in posts_raw:
+            # there are only two top level keys here; kind (comment / post) and
+            # data. No reason to keep the kind because we're only pulling posts.
+            item = item['data']
+            if not item['is_self']:
+                trimmed_links.append({
+                    'subreddit': item['subreddit'],
+                    'name': item['name'],  # remember, this is the ID: t3_8swl2n
+                    'title': item['title'],
+                    'permalink': item['permalink'],
+                    'is_nsfw': item['over_18'],
+                    'domain': item['domain'],
+                    'ups': item['ups'],
+                    'locked': item['locked'],
+                    'archived': item['archived'],
+                    'author': item.get('author', None)
+                })
+        return trimmed_links
+
+    headers = {
+        'User-Agent': generate_user_agent()
+    }
+    url = 'https://www.reddit.com/r/{}/new/.json'.format(sub)
+    result = requests.get(url, headers=headers).json()
+    # we have two states here: one has the data we want and the other is an
+    # error state. The error state looks like this:
+    # {'message': 'Too Many Requests', 'error': 429}
+
+    if result.get('error', None):
+        logging.warning('hit error state for {}'.format(sub))
+        return []
+    return parse_json_posts(result)
+
+
+def threaded_check_submissions(config):
+    """
+    Single threaded PRAW performance:
+    finished in 56.75446701049805s
+
+    Single threaded json performance:
+    finished in 16.70485234260559s
+
+    multi-threaded json performance:
+    finished in 1.3632569313049316s
+    """
+    subreddits = config.subreddits_to_check
+
+    total_posts = list()
+    # by not specifying a maximum number of threads, ThreadPoolExecutor will
+    # grab the CPU count of the current machine and multiply it by 5, allowing
+    # us to keep sane limits wherever we're running.
+    with ThreadPoolExecutor() as executor:
+        jobs = list()
+        for sub in subreddits:
+            jobs.append(executor.submit(get_subreddit_posts, sub))
+        for f in as_completed(jobs):
+            try:
+                data = f.result()
+                total_posts += data
+            except Exception as exc:
+                logging.warning('an exception was generated: {}'.format(exc))
+
+    for item in total_posts:
```

Comment:
Is there a reason why we're waiting for all threads to complete before processing the posts? Wouldn't it be nicer and more efficient to do this as soon as the jobs finish?

Also, I think it'd probably be nicer to move the domain check  into `process_post` as other filtering also occurs there. Maybe even move the entire filter process into its own method?

Reply:
I like the idea to move the filtering to its own method -- that'd probably make it clearer.

As for doing this as soon as the jobs finish, that's definitely possible but it's more complicated. Doing it this way is the recommended way in the Python3 documentation for guaranteeing that all the threads get cleaned up properly. Less work on our side.